Fix for throttle acquisition on LocoNet

pabender opened the pull request:
This addresses #4677 #4541 and perhaps others.

There was an observation made on the users list that on systems impacted by the broken LocoNet throttle acquisition process if you tried to acquire the locomotive twice, it worked the second time.

Looking at what differs in the system between the two acquisition attempts, the only difference found was that when the throttle is released after the first attempt, the slot is set to common.

This PR sets the slot to idle prior to null move that sets the slot to in use, and is basically a 1 line solution to the problem.

This was tested using a DB150 that was previously experiencing the issue.  

I've also tested with a Roco Z21 connected via a LocoBuffer (yes, it has a LocoNet port, in addition to XPressNet and a few others...) It did not experience any throttle acquisition errors either before or after the change.

I'd welcome more testing prior to merging.

rhwood commented:
Travis CI GUI failed on known threading issue with InstanceManager and was restarted.

rhwood commented:
@pabender I don't know the LocoNet well enough to comment on this, and am not in a position to test soon, but there is nothing I would question in the code, so I've removed myself as a reviewer for this.

Pugwash1 commented:
No. This is a band aid to mask the underlying problems.
If you are going to mimic a throttle then you do ask for slot, get slot, null move, write slot (for throttle ID and speed steps). With correct timings as command stations react at different speeds. If you mimic a computer you scan slots, write slot with above info and status in use. What you don't do, which happening now, is ask for slot, write slot, null move. And make sure you get acknowledgement between steps. It's very slow on older command stations. If you want the proof remove the current write slot before create throttle.

Pugwash1 commented:
The technical reason for not writing the slot before the null move is d7_SL_PURGE in stat1. It's not seen in the message. When the slot is written it gets destroyed, the null move sees the slot allocated has been updated since the ask for slot request, and ignores the null move.

pabender commented:
@Pugwash1,

So you are saying this line

https://github.com/JMRI/JMRI/blob/c4c93f6e9679c6fe28f81e35e794bb2ac54e96c7/java/src/jmri/jmrix/loconet/LnThrottleManager.java#L160

Needs to move ( logically ) after this block:

https://github.com/JMRI/JMRI/blob/c4c93f6e9679c6fe28f81e35e794bb2ac54e96c7/java/src/jmri/jmrix/loconet/LocoNetThrottle.java#L49-L54

Pugwash1 commented:
@pabender Yes, logically. The move takes a while so the slot write may get lost. Ive often wondered why the slot move was in the throttle code, not the aquisition code. ps there is also errors in the the way we dispose of throttles, it always , i think, send it idle, so locos stop, an if part of universal consist. Got to go, spouse waiting for me..

bobjacobsen commented:
The thing that confuses me about all this (and the reason I've pretty much stayed out of it) is that AFAIK LocoNet throttle acquisition worked without significant outcry from sometime in 2008 or so (when a race condition when sharing with real throttles was finally fixed) until sometime around last summer.  I'm pretty sure it wasn't in compliance with various LocoNet docs, because it was copied from how DT100 and DT400 throttles of that era worked.  But I remember it as stable.

What's changed?

devel-bobm commented:
@bobjacobsen,
My changes to add the write of ThrottleID was apparantly _one_ culprit.  It apparantly "broke" things.  But it actually helped out in at least some cases for DCS240 with "Expanded slots enabled".

There is also the problem where a "saved throttle arrangement" would blast a bunch of throttle requests upon load, and the multiple requests were "overlapped" rather than metered, resulting in multiple JMRI throttles controlling the same slot, rather than multiple throttles controlling unique slots.  I don't recall if this was an aspect of my ThrottleID changes or if it was an independent change.

pabender commented:
My primary concern with all of this is making sure the slots are stable enough that the consist tool continues to work as it did before any of @devel-bobm 's changes from last summer.

As discussed in #4677, this becomes a throttle issue because the throttle is used to set the speed and direction ( and to obtain the slots for linking ).  The discussion in #4677 includes a Loconet trace that shows the null slot move fails when the slots were previously empty.

Frankly, if moving the set throttleID after the null slot move fixes that, that solves my problem.  ( This is an embodiment of the test driven development process which says develop a test that fails, and make the smallest number of changes required to make it pass.  The tests here just require hardware to function... )

I know there are other issues related to all this, so I want to get this right, but I also know it has been several months since there was any progress made, and fixing my test case is a step in the right direction, even if it doesn't address all those other issues.

Incidentally, another issue that needs to be addressed is what happens when the release button is pressed.  The LnThrottleManager is now setting the slot to common when that happens, so the slot is no longer in use... but if we have a script or some other process ( such as the consist manager ) manipulating the specified slot, it needs to stay in use.  I believe this was the behavior prior to last summer ( though I have not verified that in the code ).

Pugwash1 commented:
Sorry, took so long to re-enable brain after interrupt  my edit crossed everything.
Yes the better than band aid is #4293, which tries to queue every thing and not interlace the requests.

Pugwash1 commented:
In https://github.com/Pugwash1/JMRI.git/GetBySlot is a small hack to demonstrate acquisitions by manipulating slots. Its just a hack, so little validation, and no stealing. But it can get you 15 throttles in less than a seconds, all working no interleafing problems. It might give someone some ideas.
It gets the status of all slots when it starts.
Searches its slots for something suitible.
Writes the slot.
Returns the throttle.

pabender commented:
@Pugwash1 

I looked through the code there, and I see that It is still doing an update of the slot before the null move.  It does this by changing the method that updates the throttleID to also update speed, address, and status.  It just happens to be updating the slot to say that it is IN_USE, instead of IDLE like this PR does.

In other words, It's effectively using the same band-aid solution I've applied here, plus some other code to search for slots that will only work on a DCS100 (and DCS200?) due to hard coding 120 as the number of slots to search through.

Pugwash1 commented:
Unless i pushed the wrong version loconetthrottle has the null move removed. I will double check.
Steve G.

On January 7, 2018 2:36:22 PM EST, Paul Bender <notifications@github.com> wrote:
>@Pugwash1 
>
>I looked through the code there, and I see that It is still doing an
>update of the slot before the null move.  It does this by changing the
>method that updates the throttleID to also update speed, address, and
>status.  It just happens to be updating the slot to say that it is
>IN_USE, instead of IDLE like this PR does.
>
>In other words, It's effectively using the same band-aid solution I've
>applied here, plus some other code to search for slots that will only
>work on a DCS100 (and DCS200?) due to hard coding 120 as the number of
>slots to search through.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

Pugwash1 commented:
The branch is getbyslot. No null move. Its not there.
Steve G.

On January 7, 2018 2:36:22 PM EST, Paul Bender <notifications@github.com> wrote:
>@Pugwash1 
>
>I looked through the code there, and I see that It is still doing an
>update of the slot before the null move.  It does this by changing the
>method that updates the throttleID to also update speed, address, and
>status.  It just happens to be updating the slot to say that it is
>IN_USE, instead of IDLE like this PR does.
>
>In other words, It's effectively using the same band-aid solution I've
>applied here, plus some other code to search for slots that will only
>work on a DCS100 (and DCS200?) due to hard coding 120 as the number of
>slots to search through.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

Pugwash1 commented:
As to the number of slots that is just a number, its a demo, not intended for production, obviously there many things wrong with it to be used as a solution.

pabender commented:
On 01/07/2018 03:02 PM, Steve G wrote:
> The branch is getbyslot. No null move. Its not there.

That is a problem.

The null move is required according to the LocoNet PE spec (in the 
paragraph spanning the bottom of page 6, top of page 7.)

pabender commented:
Ok, so doing some more work here.

I moved the setting of the throttle ID into the throttle... instead of putting it into the constructor however, 

I moved it into the notifyChangedSlot method, so the sequence becomes:

1. (manager) Request Slot
2. (manager) receive slot result from cs
3. (manager creates throttle)
4. (throttle) Null Slot move
5. (throttle) receive updated slot info from cs
6. (throttle) set Throttle ID

with the setting of the throttle id protected by a flag, so it only happens once.

but here's what happens on my DB150 when I try to acquire a throttle:

> [BF 2B 28 43]  Request slot for loco address 5544.
> [81 7E]  Master is busy.
> [81 7E]  Master is busy.
> [E7 0E 04 03 28 00 00 07 00 2B 00 00 00 15]  Report of slot 4 information:
> 	Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
> 	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
> 	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).
> [BA 04 04 45]  Set status of slot 4 to IN_USE.
> [E7 0E 04 33 28 00 00 07 00 2B 00 00 00 25]  Report of slot 4 information:
> 	Loco 5544 is Not Consisted, In-Use, operating in 128 SS mode, and is moving Forward at speed 0,
> 	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
> 	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

So all that above is steps 1-5, and that's what we should see (that is what we used to see... and what the LocoNet PE spec says we SHOULD be seeing)

Now, where we have a problem is when we try to set the throttle ID:

> [EF 0E 04 03 28 00 00 07 00 2B 00 71 02 6E]  Write slot 4 information:
> 	Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
> 	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
> 	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x02 0x71 (369).

The throttle ID is set, but the slot (which should be authoritative here, since it triggers the update) doesn't have the updated information from the command station.

Incidentally, the LocoNet PE never says ANYTHING about throttle IDs.  I know that isn't all there is to LocoNet, but it's the only authoritative source we have.

I'm going to go ahead and push this up so we can discuss those changes, with the writing of the throttle ID disabled (not removed, just disabled ) because that's the only way this code produces correct results on my DB150.

devel-bobm commented:
@pabender , 
```[EF 0E 04 **03** 28 00 00 07 00 2B 00 71 02 6E] Write slot 4 information:
Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x02 0x71 (369).```

This `OPC_WR_SL_DATA` message seems to change the slot status from "IN_USE" (from the previous `OPC_SL_RD_DATA`) to "FREE", from the bolded byte.  `OPC_SLOT_STAT1` is the "normal" way to update a slot's status; I don't know if the slot will actually take such a "slot status write" via a slot write.  But it could "drop" the slot from controlling the loco.

FYI: The LocoNet PE document does not discuss updating the slot with the "Throttle ID" information.  But it does mention <ID1> and <ID2> (a.k.a. "Throttle ID", for some values of STAT2) and gives some ambiguous info about their encoding and usage.  See below `OPC_SL_RD_DATA`  and the text describing "SLOT STATUS2" and "EXPANSION RESERVED ID1" and "EXPANSION RESERVED ID2".

devel-bobm commented:
@pabender ,

FYI - here is the (annotated) traffic from a DT500 (with expanded slots DISABLED) and DCS210 (freshly-factory-resetted) when dialing up engine 5544:

try to find address "0" (the default when a digit hasn't been pressed yet)
```
[BF 00 00 40]  Request slot for loco address 0 (short).
[E7 0E 01 03 00 00 00 47 00 00 00 00 00 53]  Report of slot 1 information:
	Loco 0 (short) is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Press "5" - throttle looks up address "5":
```
[BF 00 05 45]  Request slot for loco address 5 (short).
[E7 0E 01 03 05 00 00 47 00 00 00 00 00 56]  Report of slot 1 information:
	Loco 5 (short) is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Press "5" again, throttle tries to look up address "55":
```
[BF 00 37 77]  Request slot for loco address 55 (short).
[E7 0E 01 03 37 00 00 47 00 00 00 00 00 64]  Report of slot 1 information:
	Loco 55 (short) is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Press "4", throttle tries to look up address "554":
```
[BF 04 2A 6E]  Request slot for loco address 554.
[E7 0E 01 03 2A 00 00 47 00 04 00 00 00 7D]  Report of slot 1 information:
	Loco 554 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Press "4" again, throttle tries to look up address "5544":
```
[BF 2B 28 43]  Request slot for loco address 5544.
[E7 0E 01 03 28 00 00 47 00 2B 00 00 00 50]  Report of slot 1 information:
	Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Press "loco" to acquire address "5544"; throttle gets the slot again (presumably because there may have been other LocoNet activity which "upsets" the slot info since the last button was pressed):
```
[BF 2B 28 43]  Request slot for loco address 5544.
[E7 0E 01 03 28 00 00 47 00 2B 00 00 00 50]  Report of slot 1 information:
	Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Throttle attempts to make the slot usable to the throttle by setting "slot status" to "in-use":
```
[BA 01 01 45]  Set status of slot 1 to IN_USE.
[E7 0E 01 33 28 00 00 47 00 2B 00 00 00 60]  Report of slot 1 information:
	Loco 5544 is Not Consisted, In-Use, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).

```

Throttle marks the slot with its "throttle ID":

```
[EF 0E 01 33 28 00 00 47 00 2B 00 17 53 2C]  Write slot 1 information:
	Loco 5544 is Not Consisted, In-Use, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x53 0x17 (10647).
[B4 6F 7F 5B]  LONG_ACK: Function not implemented, no reply will follow.

```

Now the loco may be used by the throttle:
`[A0 01 00 5E]  Set speed of loco in slot 1 to 0.`

Note that the "Throttle ID write" did not attempt to change the `<STAT>` byte, and that `<SS2>` was left at 0, meaning (as best as I can interpret LocoNet PE 1.0):

- the slot does not suppress advanced consisting (`<SS2.0>` = 0).
- the '<ID1>' and `<ID2>` bytes can have "ID" usage (i.e. '<SS2.2>' = 0).
- the `<ID1>` and `<ID2>` bytes are not an "expansion", but (might?) be an `ENCODED alias` (i.e. `<SS2.3>` = 0).

devel-bobm commented:
@pabender, you wrote:

> The null move is required according to the LocoNet PE spec (in the paragraph spanning the bottom of page 6, top of page 7.)

Note that the "Loco Acquisition" sequence I posted above did NOT show an "OPC_NULL_MOVE" operation!  So the LocoNet PE 1.0 "spec" is not honored "to-the-letter" by the DT500 when working with a DCS210 with an empty slot table!

devel-bobm commented:
Please ignore my previous (now deleted) comment about about a non-present NULL_MOVE operation.  It's there...

pabender commented:
> On Jan 7, 2018, at 6:26 PM, Bob Milhaupt <notifications@github.com> wrote:
> 
> @pabender ,
> 
> Loco 5544 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
> F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
> Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x02 0x71 (369).```
> 

> This `OPC_WR_SL_DATA` message seems to change the slot status from "IN_USE" (from the previous `OPC_SL_RD_DATA`) to "FREE", from the bolded byte.  `OPC_SLOT_STAT1` is the "normal" way to update a slot's status; I don't know if the slot will actually take such a "slot status write" via a slot write.  But it could "drop" the slot from controlling the loco.

That message is my concern as well, because it is being produced inside the notifyChangedSlot method of the throttle,  but it isn't using the values that were just received from the command station.  I think it should be.

The code that produces this is:

network.sendLocoNetMessage(slot.writeThrottleID(throttleManager.getThrottleID()));

And was moved from the previous location it was called from inside the throttle manager. ( it is that line 160 reference from one of my previous posts on this PR ).  It doesn't work in the previous location, and it doesn't work here either, so either the slot's writeThrottleID needs to be re-examined, or the call completely removed.

pabender commented:
@devel-bobm 

So it looks like my current sequence matches what that dt500 does, except the writing of the slot data with the ID is broken.

I haven't changed that particular piece of code, only re-ordered its location in the sequence.

devel-bobm commented:
I infer from the symptom that the `OPC_SL_READ_DATA` that follows the `NULL_MOVE` fails to actually updates the LnSlot's `stat` variable - which is used when creating `<STAT>` in the Throttle ID message.

So, either the `LnSlot.setSlot` method is not called when the reply to the `NULL_MOVE` occurs, or the execution of `LnSlot.setSlot` is broken.  Try putting a log.info() message in the 'LnSlot.setSlot()` method, to see if it is invoked immediately after reception of the reply to the `NULL_MOVE`.

In my (close-to-JMRI/JMRI/head) codebase, I see the LnSlot.setSlot() method being invoked account the OPC_SL_READ_DATA`, whether from a JMRI throttle or when a DT500 is operated.  And I see `stat` properly updated for both.  I do not know how far my SlotManager codebase has veered from JMRI/JMRI/head, though, but I believe that they are unrelated changes.

pabender commented:
I'll check that in a little bit.

I'm also going through the process of actually adding tests that mean something to the LocoNetSlotTest class.  I already found one bug during this process that made the LocoNetSlot(LocoNetMessage) constructor fail.

pabender commented:
So the slot was being updated after the move, but the slot change message triggering setting the slot's ThrottleID was happening before the move was finished.

That was fixed by setting the throttle's slot change listener to only trigger when the slot showed in use, and now the sequence looks like this:
```

[BF 4C 25 29]  Request slot for loco address 9765.
[81 7E]  Master is busy.
[81 7E]  Master is busy.
[E7 0E 04 03 25 00 00 07 00 4C 00 00 00 7F]  Report of slot 4 information:
	Loco 9765 is Not Consisted, Free, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).
[BA 04 04 45]  Set status of slot 4 to IN_USE.
[E7 0E 04 33 25 00 00 07 00 4C 00 00 00 4F]  Report of slot 4 information:
	Loco 9765 is Not Consisted, In-Use, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x00 0x00 (0).
[EF 0E 04 33 25 00 00 07 00 4C 00 71 02 34]  Write slot 4 information:
	Loco 9765 is Not Consisted, In-Use, operating in 128 SS mode, and is moving Forward at speed 0,
	F0=Off, F1=Off, F2=Off, F3=Off, F4=Off, F5=Off, F6=Off, F7=Off, F8=Off
	Master supports LocoNet 1.1; Track Status: On/Running; Programming Track Status: Available; STAT2=0x00, ThrottleID=0x02 0x71 (369).
[B4 6F 7F 5B]  LONG_ACK: Function not implemented, no reply will follow.
```

And at that point, the slot shows in use and the throttle ID is set.

And except for a minor issue with the lead locomotive in a Command Station Consist, there isn't an issue with the consisting tool (one of my recent changes propagated an error to the tool if you tried to link/unlink a single slot, but if there is only one locomotive, there is an attempt to link which shouldn't happen... but that isn't a loconet traffic issue.)

devel-bobm commented:
Does this solution include a fix for the problem with "saved throttle layouts" performing multiple throttle acquisitions in parallel, resulting in multiple throttles pointing at the same slot, even though they think they are controlling different locos?  I don't remember if that's been addressed in JMRI/JMRI/master.

devel-bobm commented:
@pabender , Thanks for taking on this issue!

pabender commented:
@devel-bobm This does not include a fix for opening multiple throttles simultaneously.

That is a change that should be limited to the Throttle manager.  It needs to queue incoming requests if it can't handle multiple simultaneously without issues.  I can look into that, but I consider it a separate issue.

pabender commented:
travis headless restarted after

[INFO] Running jmri.configurexml.StoreXmlUserActionTest
ERROR - Proceeding to initialize class jmri.implementation.SignalSpeedMap but initialization is marked as complete [main] jmri.InstanceManager.?()
java.lang.Exception: Thread "AWT-EventQueue-0"
	at jmri.InstanceManager.getNullableDefault(InstanceManager.java:255)
	at jmri.InstanceManager.getDefault(InstanceManager.java:204)
	at jmri.implementation.DefaultSignalMastLogic.setMastAppearance(DefaultSignalMastLogic.java:913)
	at jmri.implementation.DefaultSignalMastLogic.lambda$setSignalAppearance$0(DefaultSignalMastLogic.java:785)
	at jmri.util.ThreadingUtil.lambda$runOnGUIDelayed$0(ThreadingUtil.java:165)
	at javax.swing.Timer.fireActionPerformed(Timer.java:313)
	at javax.swing.Timer$DoPostEvent.run(Timer.java:245)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

pabender commented:
@devel-bobm 

I couldn't just leave the multiple simultaneous requests hanging.

I just pushed up a change to the LnThrottleManager that queues the requests.... as soon as I go back and fix the other LocoNet throttle managers (that I forgot about) this will should do what we want it to.

I tested by creating a throttle layout with 6 throttles.  prior to adding the queue, I received 1 working slot and one that was not marked in use.  After adding the queue, I received 6 throttles marked as in use.... not bad for less than an hours work.

pabender commented:
Except for perhaps the UhlenbrockLnThrottleManager, it looks like the others are ok.

The UhlenbrockLnThrottleManager seems to duplicate a lot of code from the LnThrottleManager.

pabender commented:
Appveyor restarted after

    [junit] Testcase: testSetShowHelpBar(jmri.jmrit.display.layoutEditor.LayoutEditorTest):	Caused an ERROR
    [junit] getShowHelpBar
    [junit] java.lang.AssertionError: getShowHelpBar
    [junit] 	at jmri.jmrit.display.layoutEditor.LayoutEditorTest.testSetShowHelpBar(LayoutEditorTest.java:396)

pabender commented:
Travis headed restarted after a timeout.  The last log entry was

[INFO] Running jmri.jmrit.operations.ExceptionDisplayFrameTest

devel-bobm commented:
@pabender , Great work!

devel-bobm commented:
Restarting Appveyor headed account failure summary:
```
[ERROR] Errors: 
[ERROR]   MergePromptTest.testCTor:35->Object.wait:502->Object.wait:-2 » TestTimedOut te...
[INFO] 
[ERROR] Tests run: 32424, Failures: 0, Errors: 1, Skipped: 176
```
and the following failure details:
```[INFO] Running jmri.jmrit.logix.MergePromptTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 10.014 s <<< FAILURE! - in jmri.jmrit.logix.MergePromptTest
[ERROR] testCTor(jmri.jmrit.logix.MergePromptTest)  Time elapsed: 10.014 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 10 seconds
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:257)
	at java.awt.Dialog.show(Dialog.java:1084)
	at java.awt.Component.show(Component.java:1671)
	at java.awt.Component.setVisible(Component.java:1623)
	at java.awt.Window.setVisible(Window.java:1014)
	at java.awt.Dialog.setVisible(Dialog.java:1005)
	at jmri.jmrit.logix.MergePrompt.<init>(MergePrompt.java:148)
	at jmri.jmrit.logix.MergePromptTest.testCTor(MergePromptTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

pabender commented:
Headed Travis timed out, apparently during one of the cucumber tests.

pabender commented:
Travis passed now, so just waiting for coveralls.  

@devel-bobm @Pugwash1 @bobjacobsen do you have any objection to this getting merged now?

devel-bobm commented:
@pabender , I'm ok with a merge of this code.

Pugwash1 commented:
@devel-bobm @Pugwash1 @bobjacobsen I have no objections. In my testing it got throttles correctly via wifi and itself. It didnt get all the throttles on a layout, but more than it used to, but that is high hanging, no much used, fruit. But, it is a large improvement in what we have, and it provides a way forward.  Sorry to be so tardy.
Steve G.

devel-bobm commented:
@pugwash1 - I would be interested in any details you can provide about 

>  It didnt get all the throttles on a layout, but more than it used to, but that is high hanging, no much used, fruit.

Apparantly, the current code does not handle one or more "exception" cases, and we should try to understand and address those.  I would think that a separate issue would be in order, and perhaps such reporting should wait until after this has been merged to "master".

Pugwash1 commented:
@devel-bobm I agree. I was trying to work out what was happening, and will continue to do so, but the clock is ticking and we need the basic fix urgently and this (loading throttle layouts) is not a show stopper. The auto start auto dispatched trains are loaded at 5 a second which solved a few more than this problem.
Steve Gigiel

pabender commented:
Ok, I am going to go ahead and merge this.

@Pugwash1 can you open an issue for what you are seeing, including logs ( both Loconet and console ) so I can see what is happening, and please include the throttle layout information, or even better, the zipped profile, so I can try to replicate.
